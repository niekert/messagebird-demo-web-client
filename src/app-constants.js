// Status codes
export const INITIAL = 'INITIAL';
export const PENDING = 'PENDING';
export const OK = 'OK';
export const ERROR = 'ERROR';
