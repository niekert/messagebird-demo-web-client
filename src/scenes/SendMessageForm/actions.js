import * as messagesApi from 'api/messages';

export const SEND_MESSAGE = 'SEND_MESSAGE';
export const SEND_MESSAGE_ERROR = 'SEND_MESSAGE_ERROR';
export const SEND_MESSAGE_SUCCESS = 'SEND_MESSAGE_SUCCESS';

/**
 * Send a text message with the given paramters
 * @param {String} recipient - Phone number(s) of the recipient, comma separated
 * @param {String} originator - Name of the originator
 * @param {String} body - Body of the text
 * @returns {Function} thunk
 */
export function sendMessage(recipient, originator, body) {
  return async dispatch => {
    dispatch({ type: SEND_MESSAGE });

    try {
      const message = await messagesApi.sendMessage(
        recipient.split(','),
        originator,
        body,
      );

      console.log('message is ', message);
    } catch (error) {
      window.error = error;
      console.log('error', error);
      dispatch({
        type: SEND_MESSAGE_ERROR,
        payload: {
          error,
        },
      });
    }
  };
}
