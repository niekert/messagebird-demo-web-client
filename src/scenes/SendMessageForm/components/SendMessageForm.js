import React from 'react';
import { func, instanceOf } from 'prop-types';
import { OK, ERROR, PENDING } from 'app-constants';
import { status as statusPropType } from 'customPropTypes';
import styled from 'styled-components';
import { H1 } from 'style/Headers';
import { smallUp } from 'style/breakpoints';
import LoadingButton from 'shared/components/LoadingButton';
import { Form, TextArea, SubmitButton, PhoneInput } from 'style/Form';
import { ContentCard } from 'style/Cards';
import SuccessBanner from './SuccessBanner';
import ErrorBanner from './ErrorBanner';

/**
 * Super crappy way of formatting a phone number to a normal string
 * Should be done in a decent way in a real app
 * @param {String} number - The phone number to format
 */
function formatPhoneNumber(phoneNumber) {
  return phoneNumber.replace(/\s/g, '').replace('+', '00');
}

const Wrapper = styled.div`
  ${smallUp`
    position: sticky;
    top: 50px;
  `};
`;

class SendMessageForm extends React.Component {
  static propTypes = {
    status: statusPropType.isRequired,
    error: instanceOf(Error),
    sendMessage: func.isRequired,
  };

  static defaultProps = {
    error: null,
  };

  state = {
    recipient: '',
    body: '',
  };

  componentWillReceiveProps(nextProps) {
    if (nextProps.status === OK && this.props.status !== OK) {
      // Reset the form if all is GOOD
      this.setState({
        recipient: '',
        body: '',
      });
    }
  }

  handleInputChange = e => {
    const { name, value } = e.target;
    this.setState({
      [name]: value,
    });
  };

  handleChange = phoneNumber => {
    this.setState({
      recipient: phoneNumber,
    });
  };

  handleSubmit = e => {
    e.preventDefault();
    const { recipient, body } = this.state;
    const formattedRecipient = formatPhoneNumber(recipient);

    // TODO: Could do some additional validation before calling the action

    this.props.sendMessage(formattedRecipient, body);
  };

  render() {
    const { status, error } = this.props;
    const { recipient, body } = this.state;

    return (
      <Wrapper>
        <H1>Send a message</H1>
        <ContentCard>
          {status === OK && <SuccessBanner />}
          {status === ERROR &&
            !!error && <ErrorBanner message={error.toString()} />}
          <Form onSubmit={this.handleSubmit}>
            <PhoneInput
              name="recipient"
              defaultCountry="nl"
              onChange={this.handleChange}
              value={recipient}
            />
            <TextArea
              placeholder="Body"
              name="body"
              required
              value={body}
              onChange={this.handleInputChange}
            />
            <LoadingButton
              component={SubmitButton}
              isLoading={status === PENDING}
            >
              Send message
            </LoadingButton>
          </Form>
        </ContentCard>
      </Wrapper>
    );
  }
}

export default SendMessageForm;
