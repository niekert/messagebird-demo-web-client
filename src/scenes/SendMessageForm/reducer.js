import { createReducer } from 'redux-create-reducer';
import { INITIAL, PENDING, OK, ERROR } from 'app-constants';
import {
  SEND_MESSAGE,
  SEND_MESSAGE_ERROR,
  SEND_MESSAGE_SUCCESS,
} from './actions';

const initialState = {
  status: INITIAL,
  error: null,
};

export default createReducer(initialState, {
  [SEND_MESSAGE]() {
    return {
      status: PENDING,
      error: null,
    };
  },
  [SEND_MESSAGE_SUCCESS]() {
    return {
      status: OK,
      error: null,
    };
  },
  [SEND_MESSAGE_ERROR](state, action) {
    return {
      status: ERROR,
      error: action.payload.error,
    };
  },
});
