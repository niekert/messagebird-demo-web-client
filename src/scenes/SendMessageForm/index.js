import React from 'react';
import { connect } from 'react-redux';
import { sendMessage } from './actions';
import SendMessageForm from './components/SendMessageForm';

const mapStateToProps = state => state;

// TODO: Implement container logic for SendMessageForm component

export default connect(mapStateToProps, { sendMessage })(SendMessageForm);
