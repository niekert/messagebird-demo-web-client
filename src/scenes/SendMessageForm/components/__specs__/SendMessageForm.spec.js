import React from 'react';
import { INITIAL } from 'app-constants';
import { shallow } from 'enzyme';
import SendMessageForm from '../SendMessageForm';

describe('<SendMessageForm />', () => {
  let props;
  beforeEach(() => {
    props = {
      status: INITIAL,
      sendMessage: jest.fn(),
    };
  });

  test('renders correctly', () => {
    const wrapper = shallow(<SendMessageForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('Sends a message on submit', () => {
    const wrapper = shallow(<SendMessageForm {...props} />);
    wrapper.setState({ recipient: '0634322664', body: 'nice message wow' });

    wrapper.instance().handleSubmit({ preventDefault: () => {} });

    expect(props.sendMessage).toHaveBeenCalledTimes(1);
    expect(props.sendMessage).toHaveBeenLastCalledWith(
      '0634322664',
      'nice message wow',
    );
  });
});
