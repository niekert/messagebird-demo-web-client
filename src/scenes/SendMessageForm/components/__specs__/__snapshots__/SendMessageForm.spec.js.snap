// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SendMessageForm /> renders correctly 1`] = `
<SendMessageForm__Wrapper>
  <Headers__H1>
    Send a message
  </Headers__H1>
  <Cards__ContentCard>
    <Form
      onSubmit={[Function]}
    >
      <Form__Input
        defaultCountry="nl"
        name="recipient"
        onChange={[Function]}
        value=""
      />
      <Form__Input
        name="body"
        onChange={[Function]}
        placeholder="Body"
        required={true}
        value=""
      />
      <LoadingButton
        component={[Function]}
        isLoading={false}
      >
        Send message
      </LoadingButton>
    </Form>
  </Cards__ContentCard>
</SendMessageForm__Wrapper>
`;
